### PR TITLE
persist: another special case in apply_diffs_spine

### PR DESCRIPTION
**Pull request by danhhz**
Sniff out a newly inserted batch (one whose lower lines up with the current upper) and handle that now. Then fall through to the rest of the handling on whatever is left.

Touches #15493

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Review by pH14 — APPROVED** [review]
LGTM -- the overall complexity in our read repair is a little out of control, but this particular change seems reasonable + "straightforward". The reason for the change is very subtle, but preemptively pushing the `spine.upper == batch.lower` batch seems safe to me

File: src/persist-client/src/internal/state_diff.rs [hunk 1]
```diff
@@ -44,8 +44,11 @@ pub enum StateFieldValDiff<V> {
 #[derive(Clone, Debug)]
 #[cfg_attr(any(test, debug_assertions), derive(PartialEq))]
 pub struct StateFieldDiff<K, V> {
-    pub key: K,
+    // These are intentionally in an unnatural order because in the cases we've
+    // seen in the wild, it's been more useful to have the val printed first in
+    // the std::fmt::Debug impl.
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-client/src/internal/state_diff.rs]: done!

File: src/persist-client/src/internal/state_diff.rs [hunk 2]
```diff
@@ -460,24 +463,28 @@ fn diff_field_spine<T: Timestamp + Lattice>(
...
-            key,
-            val: StateFieldValDiff::Insert(()),
-        }] => {
-            // Ignore merge_reqs because whichever process generated this diff is
-            // assigned the work.
-            let _merge_reqs = trace.push_batch(key.clone());
+    // Another special case: sniff out a newly inserted batch (one whose lower
+    // lines up with the current upper) and handle that now. Then fall through
+    // to the rest of the handling on whatever is left.
+    if let Some(insert) = sniff_insert(&mut diffs, trace.upper()) {
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist-client/src/internal/state_diff.rs:480]: without thinking about it _too_ hard, I think maybe this already does that? the sniff_insert will handle it via the `push_batch` "fast path" and if the remaining diffs e.g. cleanly work for the `apply_merge_res` "fast path" that will also happen. in practice, the metrics will show if this changed the ratio when it gets deployed

**Review by danhhz — COMMENTED** [review]
TFTR!

**Pull request merged**